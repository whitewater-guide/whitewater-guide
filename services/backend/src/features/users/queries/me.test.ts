import { holdTransaction, rollbackTransaction } from '@db';
import { redis } from '@redis';
import {
  ADMIN,
  BOOM_USER_3500,
  BOOM_USER_3500_ID,
  EDITOR_GA_EC,
  TEST_USER,
  TEST_USER_ID,
} from '@seeds/01_users';
import { GROUP_ALL } from '@seeds/03_groups';
import { REGION_ECUADOR, REGION_GEORGIA } from '@seeds/04_regions';
import { anonContext, fakeContext, runQuery } from '@test';
import axios from 'axios';

jest.mock('axios');

beforeEach(async () => {
  await holdTransaction();
  jest.resetAllMocks();
  await redis.flushall();
  (axios.get as any).mockResolvedValue({
    data: { data: { url: 'https://ww.guide/pic.jpg' } },
  });
});
afterEach(async () => {
  await rollbackTransaction();
  redis.removeAllListeners();
});

const query = `
{
  me {
    id
    name
    avatar
    admin
    email
    createdAt
    updatedAt
    language
    imperial
    editorSettings {
      language
    }
    accounts {
      id
      provider
    }
  }
}
`;

it('should return null for anon', async () => {
  const result = await runQuery(query, undefined, anonContext());
  expect(result.errors).toBeUndefined();
  expect(result.data!.me).toBeNull();
});

it('should return local user', async () => {
  const result = await runQuery(query, undefined, fakeContext(EDITOR_GA_EC));
  expect(result.errors).toBeUndefined();
  expect(result.data!.me).toMatchSnapshot();
});

it('should return facebook user', async () => {
  const result = await runQuery(query, undefined, fakeContext(ADMIN));
  expect(result.errors).toBeUndefined();
  expect(result.data!.me).toMatchSnapshot();
});

it('should cache fb avatar', async () => {
  await runQuery(query, undefined, fakeContext(ADMIN));
  const result = await runQuery(query, undefined, fakeContext(ADMIN));
  expect(axios.get).toHaveBeenCalledTimes(1);
  expect(result.data!.me).toMatchObject({ avatar: 'https://ww.guide/pic.jpg' });
});

it('should return purchased regions', async () => {
  const queryWithRegions = `
  {
    me {
      id
      purchasedRegions {
        id
        name
      }
    }
  }
  `;
  const result = await runQuery(
    queryWithRegions,
    undefined,
    fakeContext(TEST_USER),
  );
  expect(result.errors).toBeUndefined();
  expect(result.data!.me).toMatchObject({
    id: TEST_USER_ID,
    purchasedRegions: [
      {
        id: REGION_GEORGIA,
        name: 'Georgia',
      },
    ],
  });
});

it('should return purchased groups', async () => {
  const queryWithRegions = `
  {
    me {
      id
      purchasedGroups {
        id
        name
      }
    }
  }
  `;
  const result = await runQuery(
    queryWithRegions,
    undefined,
    fakeContext(BOOM_USER_3500),
  );
  expect(result.errors).toBeUndefined();
  expect(result.data!.me).toMatchObject({
    id: BOOM_USER_3500_ID,
    purchasedGroups: [
      {
        id: GROUP_ALL,
        name: 'All regions',
      },
    ],
  });
});

it('should not return purchased regions with invalid transactions', async () => {
  const queryWithRegions = `
  {
    me {
      id
      purchasedRegions {
        id
        name
      }
    }
  }
  `;
  const result = await runQuery(
    queryWithRegions,
    undefined,
    fakeContext(TEST_USER),
  );
  expect(result.errors).toBeUndefined();
  expect(result.data!.me.purchasedRegions).not.toContainEqual(
    expect.objectContaining({ id: REGION_ECUADOR }),
  );
});
