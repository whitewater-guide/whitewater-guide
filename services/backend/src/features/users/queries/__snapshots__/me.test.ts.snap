// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return facebook user 1`] = `
Object {
  "accounts": Array [
    Object {
      "id": "__fb_admin_id__",
      "provider": "facebook",
    },
  ],
  "admin": true,
  "avatar": "https://ww.guide/pic.jpg",
  "createdAt": "2017-02-01T00:00:00.000Z",
  "editorSettings": null,
  "email": "kaospostage@gmail.com",
  "id": "bed59990-749d-11e7-8cf7-a6006ad3dba0",
  "imperial": false,
  "language": "en",
  "name": "Ivan Ivanov",
  "updatedAt": "2017-02-01T00:00:00.000Z",
}
`;

exports[`should return local user 1`] = `
Object {
  "accounts": Array [],
  "admin": false,
  "avatar": "https://scontent.xx.fbcdn.net/v/t1.0-1/p50x50/1780622_1380926282175020_67535950_n.jpg?oh=80afda8f3e3d817e233fed522e2d9f1a&oe=5A1B4612",
  "createdAt": "2017-03-02T00:00:00.000Z",
  "editorSettings": Object {
    "language": "en",
  },
  "email": "fish.munga@yandex.ru",
  "id": "477a0bec-8a78-11e7-b3e6-9beeff45d731",
  "imperial": false,
  "language": "en",
  "name": "Konstantin Kuznetsov",
  "updatedAt": "2017-03-02T00:00:00.000Z",
}
`;
